### PR TITLE
Pttb 247 backend edit client

### PR DESCRIPTION
**Pull request by smellycat5**
Added Edit client function

### CODE REVIEW

File: app/Http/Controllers/ClientController.php [hunk 1]
```diff
@@ -37,4 +40,23 @@ public function viewAllClients()
         ], 200);
     
     }
+    public function EditClient(EditClientRequest $request, $id)
```
Comment (Amogh-svge) [inline app/Http/Controllers/ClientController.php:43]: add comments

File: app/Services/ClientService.php [hunk 2]
```diff
@@ -15,8 +15,11 @@ public static function addClient(array $validatedAddClient): bool
         return true;
     }
 
-    public static function editClient(array $validatatedEditClient)
+    public static function editClient(array $validatatedEditClient, $id)
```
Comment (Amogh-svge) [inline app/Services/ClientService.php:18]: add comments

File: app/Http/Controllers/ClientController.php [hunk 3]
```diff
@@ -7,13 +7,16 @@
...
+use Illuminate\Http\Response;
+
 
 class ClientController extends Controller
 {
-    public function addActivity(AddClientRequest $request): JsonResponse
+    public function addClient(AddClientRequest $request): JsonResponse
     {  
         $validatedAddClient = $request->validated();
         $result = ClientService::addClient($validatedAddClient);
```
Comment (Amogh-svge) [inline app/Http/Controllers/ClientController.php:22]: inject clientService dependency in constructor

Comment (Amogh-svge) [inline app/Http/Controllers/ClientController.php:22]: inject clientService dependency in constructor

File: app/Http/Controllers/ClientController.php [hunk 4]
```diff
@@ -29,12 +32,31 @@ public function addActivity(ClientRequest $request): JsonResponse
 
     public function viewAllClients()
     {
-        $projects = Client::all();
+        $clients = Client::all();
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:35]: try catch

File: app/Http/Controllers/ClientController.php [hunk 5]
```diff
@@ -29,12 +32,31 @@ public function addActivity(ClientRequest $request): JsonResponse
...
+            'total' => count($clients),
+            'clients' => ClientResource::collection($clients)
         ], 200);
     
     }
+    public function EditClient(EditClientRequest $request, $id)
+    {
+        $validatedEditClient = $request->validated();
+        try{
+            $result = ClientService::EditCLient($validatedEditClient, $id);
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:47]: use by dependency injection

File: app/Services/ClientService.php [hunk 6]
```diff
@@ -7,11 +7,19 @@
...
     {
         $log = Client::create($validatedAddClient);
 
         if (!is_object($log)) return false;
         return true;
     }
+
+    public static function editClient(array $validatatedEditClient, $id)
+    {
+        $client = Client::where('id', $id)->firstorfail();
```
Comment (sahajmalla) [inline app/Services/ClientService.php:20]: try catch

File: routes/api.php [hunk 7]
```diff
@@ -73,6 +73,7 @@
         Route::controller(ClientController::class)->prefix('client')->group(function () {
             Route::post('/', 'addActivity');
             Route::get('/', 'viewAllClients');
+            Route::patch('/{id}', 'editClient');
```
Comment (sahajmalla) [inline routes/api.php:76]: try using resourceful routes

File: app/Exceptions/Handler.php [hunk 8]
```diff
@@ -38,4 +41,17 @@ public function register()
             
         });
     }
+
+    public function render($request, Throwable $exception)
```
Comment (sahajmalla) [inline app/Exceptions/Handler.php:45]: what are you trying to do here?

File: app/Http/Controllers/ClientController.php [hunk 9]
```diff
@@ -29,12 +32,31 @@ public function addActivity(ClientRequest $request): JsonResponse
...
 
         return response()->json([
-            'total' => count($projects),
-            'clients' => ClientResource::collection($projects)
+            'total' => count($clients),
+            'clients' => ClientResource::collection($clients)
         ], 200);
     
     }
+    public function EditClient(EditClientRequest $request, $id)
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:43]: Rename the function to resourceful funtion i.e. show, index, update, destroy etc

**Review by sahajmalla — CHANGES_REQUESTED** [review]

File: app/Exceptions/Handler.php [hunk 10]
```diff
@@ -38,4 +41,17 @@ public function register()
...
+    {
+        if ($exception instanceof ModelNotFoundException) {
```
Comment (sahajmalla) [inline app/Exceptions/Handler.php:47]: remove this

File: app/Http/Controllers/ClientController.php [hunk 11]
```diff
@@ -7,34 +7,93 @@
...
 {
-    public function addActivity(ClientRequest $request): JsonResponse
-    {  
-        $validatedAddClient = $request->validated();
-        $result = ClientService::addProject($validatedAddClient);
-        if (!$result) {
+    protected Client $client;
+    protected ClientService $clientService;
+
+    public function __construct(Client $client, ClientService $clientService)
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:22]: Why are you passing Client model?

File: app/Http/Requests/EditClientRequest.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    public function rules()
+    {
+        return [
+            'client_name' => 'required|regex:^[A-Za-z]+(?:\s[A-Za-z]+)+$|max:255', //makes sure the name only accepts aplabetic characters and some specific name formats.
+            'client_number' => 'required|numeric|digits:10',
+            'status' => 'required | boolean',
+        ];
+    }
+}
```
Comment (sahajmalla) [inline app/Http/Requests/EditClientRequest.php:32]: Add one space below

File: app/Models/Client.php [hunk 13]
```diff
@@ -17,6 +17,8 @@ class Client extends Model
         'status',
     ];
 
+    protected int $perpage = 50;
```
Comment (sahajmalla) [inline app/Models/Client.php:20]: why per page 50? that is too much

File: routes/api.php [hunk 14]
```diff
@@ -71,8 +71,9 @@
             
         });
         Route::controller(ClientController::class)->prefix('client')->group(function () {
-            Route::post('/', 'addActivity');
-            Route::get('/', 'viewAllClients');
+            Route::get('/', 'index');
```
Comment (sahajmalla) [inline routes/api.php:74]: use route resource and add except show, destroy

**Review by sahajmalla — CHANGES_REQUESTED** [review]

File: app/Exceptions/Handler.php [hunk 15]
```diff
@@ -2,7 +2,10 @@
 
 namespace App\Exceptions;
 
+use Illuminate\Database\Eloquent\ModelNotFoundException;
```
Comment (sahajmalla) [inline app/Exceptions/Handler.php:5]: remove unused imports

File: app/Http/Controllers/ClientController.php (hunk 11 above)
Comment (sahajmalla) [reply app/Http/Controllers/ClientController.php:22]: Where have you used? I don't see any

File: app/Http/Controllers/ClientController.php [hunk 16]
```diff
@@ -7,34 +7,91 @@
...
+     * @return void
+     */
+    public function index()
+    {
+        try {
+            return response()->json([
+                'clients' => ClientResource::collection($this->client->paginate())
+            ], 200);
+        } catch (ModelNotFoundException $modelNotFoundException) {
+            return response()->json([
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:40]: Use the response trait  everywhere that was created by @Amogh-svge

File: app/Http/Controllers/ClientController.php [hunk 17]
```diff
@@ -7,34 +7,91 @@
...
+                'message' => 'No Clients to display',
+            ], Response::HTTP_BAD_REQUEST);
+        } catch (\Exception $exception) {
+            return response()->json([
+                'message' => $exception->getMessage(),
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:45]: Dont give exception message to the user. Add message i.e. Something went wrong or that is meaningful. Also log the exception message.

File: app/Http/Controllers/ClientController.php [hunk 18]
```diff
@@ -7,34 +7,91 @@
...
+     */
+    public function store(AddClientRequest $request): JsonResponse
+    {
+        try {
+            $validatedAddClient = $request->validated();
+            $result =  $this->clientService->addClient($validatedAddClient); ///// use dependency
+            return response()->json([
+                'message' => 'Client added successfully'
+            ]);
+        } catch (Exception $exception) {
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:64]: Try to catch what service is throwing. Eg: model not found exception here.

File: app/Http/Controllers/ClientController.php [hunk 19]
```diff
@@ -7,34 +7,91 @@
...
-        $projects = Client::all();
-
-        return response()->json([
-            'total' => count($projects),
-            'clients' => ClientResource::collection($projects)
-        ], 200);
-    
+        $validatedEditClient = $request->validated();
+        try {
+            $result =  $this->clientService->EditCLient($validatedEditClient, $id); ////// use dependency
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:82]: use camel case. Why EditCLient?

**Review by sahajmalla — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/ClientController.php (hunk 11 above)
Comment (smellycat5) [reply app/Http/Controllers/ClientController.php:22]: line number 37

File: app/Http/Controllers/ClientController.php [hunk 20]
```diff
@@ -7,32 +7,88 @@
...
+     * Undocumented function
+     *
+     * @return void
+     */
+    public function index()
+    {
+        try {
+            return $this->successResponse([ClientResource::collection($this->client->paginate())]);
+        } catch (ModelNotFoundException $modelNotFoundException) {
+            return $this->errorResponse([],'No Clients to display',Response::HTTP_BAD_REQUEST);
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:41]: Change the response code to not found not bad request

**Review by sahajmalla — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/ClientController.php [hunk 21]
```diff
@@ -2,37 +2,91 @@
...
+     */
+    public function store(AddClientRequest $request): JsonResponse
+    {
+        try {
+            $validatedAddClient = $request->validated();
+            $this->clientService->addClient($validatedAddClient);
+            return $this->successResponse([], 'Client added successfully');
+        } catch (QueryException $queryException) {
+            Log::error($queryException->getMessage());
+            return $this->errorResponse([], 'Query Exception', Response::HTTP_BAD_REQUEST);
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:61]: Give message i.e. could not add client as something is wrong with the query. Give proper message update other query exception message too.

File: app/Http/Controllers/ClientController.php [hunk 22]
```diff
@@ -2,37 +2,91 @@
...
-        $clients = Client::with(['projects'])->paginate();
-        return ClientResource::collection($clients);   
-        return response()->json([
-            'clients' => ClientResource::collection($clients)
-        ], 200);
+        try {
+            $validatedEditClient = $request->validated();
+            $result =  $this->clientService->editCLient($validatedEditClient, $client);
+            return $this->successResponse([$result], 'Client Edited successfully');
+        } catch (ModelnotFoundException $modelNotFoundException) {
```
Comment (sahajmalla) [inline app/Http/Controllers/ClientController.php:81]: Check the name of exception this is mistake

File: app/Services/ClientService.php [hunk 23]
```diff
@@ -2,16 +2,70 @@
...
+    /**
+     * Undocumented function
+     *
+     * @param array $validatedAddClient
+     * @return void
+     */
+    public static function addClient(array $validatedAddClient)
+    {
+        try {
+            Client::create($validatedAddClient);
```
Comment (sahajmalla) [inline app/Services/ClientService.php:41]: At some place you are using $clients. At some you are using model. Make it consistent

**Review by sahajmalla — CHANGES_REQUESTED** [review]

**Review by 1Aayush3 — APPROVED** [review]

**Pull request merged**